job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.job.workflow;
+
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.exception.JobDoesNotExistException;
+import alluxio.exception.status.ResourceExhaustedException;
+import alluxio.job.JobConfig;
+import alluxio.job.wire.JobInfo;
+import alluxio.job.wire.Status;
+import alluxio.job.wire.TaskInfo;
+import alluxio.job.workflow.WorkflowConfig;
+import alluxio.job.workflow.WorkflowExecution;
+import alluxio.job.workflow.WorkflowExecutionRegistry;
+import alluxio.job.wire.WorkflowInfo;
+import alluxio.master.job.JobMaster;
+
+import org.apache.commons.compress.utils.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * The {@link WorkflowTracker}.
+ */
+public class WorkflowTracker {
+  private static final Logger LOG = LoggerFactory.getLogger(WorkflowTracker.class);
+
+  private final JobMaster mJobMaster;
+
+  private final ConcurrentHashMap<Long, WorkflowExecution> mWorkflows;
+  private final ConcurrentHashMap<Long, ConcurrentHashSet<Long>> mWaitingOn;
+  private final ConcurrentHashMap<Long, ConcurrentHashSet<Long>> mChildren;
```

Comment:
I could make these HashMaps but I believe if I did, I would also have to synchronize all of the read operations as well. 

The write operations are synchronized because the code is going through multiple hashmaps and its ConcurrentHashSet. If I were a lot more careful about this coding, I may be able to remove it but I don't expect concurrency of registering new jobs to be a huge bottleneck (yet).

Reply:
Could you open an issue to track this? I'm more concerned about having maintainable / clean code.